lib/genome/groupers/gene_grouper.rb

```diff
@@ -1,160 +1,111 @@
+              newly_added_claims_count += grouped_claims.length
+              if grouped_claims.length > 0
+                add_attributes(grouped_claims)
+                add_categories(grouped_claims)
+              end
+            end
+          end
+        end until newly_added_claims_count == 0
       end
 
-      def self.create_groups
-        @alt_to_entrez.each_key do |gene_name|
-          gene_claims = @alt_to_entrez[gene_name].map(&:gene_claim)
-          gene = @gene_names_to_genes[gene_name].first if @gene_names_to_genes[gene_name]
+      def add_members(claims)
+        grouped_claims = Set.new()
+        claims.each do |gene_claim|
+
+          if (gene = DataModel::Gene.where('upper(name) = ? or upper(long_name) = ?', gene_claim.name.upcase, gene_claim.name.upcase)).any?
+            add_gene_claim_to_gene(gene_claim, gene.first)
+            grouped_claims << gene_claim
+            next
+          elsif (gene_alias = DataModel::GeneAlias.where('upper(alias) = ?', gene_claim.name.upcase)).any?
+            add_gene_claim_to_gene(gene_claim, gene_alias.first.gene)
+            grouped_claims << gene_claim
+            next
+          end
 
-          if gene
-            gene_claims.each do |gene_claim|
-              gene_claim.gene = gene if gene_claim.gene.nil?
-              gene_claim.save
+          gene_claim.gene_claim_aliases.each do |gene_claim_alias|
+            if (gene = DataModel::Gene.where('upper(name) = ? or upper(long_name) = ?', gene_claim_alias.alias.upcase, gene_claim_alias.alias.upcase)).any?
+              add_gene_claim_to_gene(gene_claim, gene.first)
+              grouped_claims << gene_claim
+              break
+            elsif (gene_alias = DataModel::GeneAlias.where('upper(alias) = ?', gene_claim_alias.alias.upcase)).any?
+              add_gene_claim_to_gene(gene_claim, gene_alias.first.gene)
+              grouped_claims << gene_claim
+              break
             end
-          else
-            @gene_names_to_genes[gene_name] = [DataModel::Gene.new.tap do |g|
-              g.name = gene_name
-              g.gene_claims = gene_claims
-              g.save
-            end]
           end
         end
+        return grouped_claims
       end
 
-      def self.add_members
-        gene_claims_not_in_groups.each do |gene_claim|
-
-          indirect_groups = []
-          direct_groups = []
-
-          direct_groups << gene_claim.name if @gene_names_to_genes[gene_claim.name]
-          gene_claim.gene_claim_aliases.map(&:alias).each do |gene_claim_alias|
-            direct_groups << gene_claim_alias if @gene_names_to_genes[gene_claim_alias]
-            alt_genes = @alt_to_other[gene_claim_alias].map(&:gene_claim)
-            alt_genes.each do |alt_gene|
-              indirect_gene = alt_gene.gene
-              indirect_groups << indirect_gene.name if indirect_gene
-            end
+      def create_gene_alias(gene_id, name, source)
```

Comment:
This is a departure from our previous direct/indirect grouping strategy.

One of the key problems with a "first match" approach like this is that it can lead to clusters of unrelated gene claims getting grouped together. Imagine if a gene claim has a mix of good and bad aliases (e.g. good: _ENSG\d+_ bad: _"N/A"_ or _"HTTP"_). It groups properly because of the good alias, but future claims group against it matching on _bad_ aliases. This could cause incorrect _good_ (but previously ungrouped) aliases to group with the wrong gene.

Previously, we handled this by requiring that a claim had only a single direct match (match to a `Gene`), and if it had no direct matches, group if it had only a single indirect match (match to a `GeneAlias`). Thus, if a bad alias got thrown in the mix, the claim would match ONLY if no other gene aliases matched it and the bad alias existed in only one other alias.

What do you think? Is that being handled here, and I'm missing it?